### ISSUE TITLE
Remove support for _count API

### ISSUE THREAD

**Issue by dadoonet** [elastic/elasticsearch#25548]
I can see that some users are confused by the fact we can count for documents by using:

* `_count` API
* `_search` API with `size: 0`

It sounds like that some options available to `_search` API can not be used in `_count` like `post_filter` as reported in https://discuss.elastic.co/t/the-new-version-of-es--count-doesnt-support-post-filter/91825 (I did not test it though).

I spoke with @jpountz about it and he explained me that the performance of using `_count` vs `_search` is the same.
So we agreed that we should open a discussion about removing the support for the `_count` API as it does not really buy us anything.

Thoughts?

**Comment by javanna** [comment]
We discussed this some time ago in #13928 and we went for making `_count` a shortcut to `_search`, but we removed the api from the transport client and all of its code that made it a different API compared to search. It was proposed to remove the `_count` endpoint too, but an objection we got was that removing `_count` from the REST layer wouldn't buy us much, while users tend to like and use such endpoint.

See also #11198, #14166 and #14165 for the changes made.

The only difference I see between `_count` and `_search` is that the former parses only the `query` part of the request body, which seems to make sense as all the other sections would require executing the search api rather than the count api.

Regarding post_filter, I do not understand why it would be useful with the `_count` api: doesn't it concern aggregation which are not part of the `_count` api either?

In conclusion: I am not against removing `_count`, but I don't think it's a big burden to maintain its REST endpoint if users find it useful. I think it does the right thing today compared to what it used to do a long time ago when it was a separate API.

**Comment by jpountz** [comment]
One thing that makes me lean towards removing it is that when there are two ways to do the same thing, users tend to wonder which one is better. So it makes the API a bit confusing. By being a bit leaner, `_count` might actually look like it is more efficient than a `_search` with size `0`.

**Comment by talabu** [comment]
Having both api that tend to do same thing and sometimes return different result could be confusing. 
I used the _count api as part of my deployment workflow and was getting disparate result compared with _search api. Later realized it was due to attachment types defined in the mapping. 
I never saw that in the documentation.

**Comment by javanna** [comment]
hi @talabu they may have returned different results in the past but they should not since 2.0 as they are effectively the same code. Do you have an example of different results from 2.0 on?

**Comment by s1monw** [comment]
we discussed this in fixit-friday and we concluded that we are in general should try to reduce the number of APIs we have. Also there should be only one way to do things but for the `_count` endpoint we think we should keep it since it's such a basic API that a significant amount of people would search for _count_ functionality that it's a low overhead to keep this syntactic sugar.

**Comment by dadoonet** [comment]
Fair enough. Should we add support for `post_filter` then? As it creates confusion.

When someone uses `post_filter` with `_search` endpoint, he can get X results but the exact same body with `_count` might give more hits.

**Comment by javanna** [comment]
I don't get why one would use post_filter with the `_count` api. I think we are good as long as `_count` throws error whenever `post_filter` is specified. The count endpoint only supports the query section of search request body, nothing else.

**Comment by talabu** [comment]
Hi @javanna. I am on version 2.3.1. _search returns the correct count. We still see much higher values returned for doc.count in _cat/indices and _stats endpoints.

**Comment by s1monw** [comment]
> We still see much higher values returned for doc.count in _cat/indices and _stats endpoints.

these APIs take deleted documents into account.

**Comment by talabu** [comment]
I have a number of **nested** field type defined in the mapping. I guess this is affecting the doc.count value returned.

**Comment by javanna** [comment]
yes @talabu if you use nested, every nested document is a separate lucene document, so lucene counts them in its stats, but the Elasticsearch APIs filter them e.g. search, count.
